File: sol/solution/0101-0200/0106/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static int p, i;
    public static TreeNode buildTree(int[] inorder, int[] postorder) {
        p = postorder.length - 1;
        i = p;
        return buildTree(inorder, postorder, Integer.MAX_VALUE);
    }
    private static TreeNode buildTree(int[] inorder, int[] postorder, int parentVal) {
        if (i < 0 || parentVal == inorder[i]) {
            return null;
        }
        int val = postorder[p];
        --p;
        TreeNode root = new TreeNode(val);
        root.right = buildTree(inorder, postorder, val);
        --i;
        root.left = buildTree(inorder, postorder, parentVal);
        return root;
    }
}

File: sol/solution/0101-0200/0106/Solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* newNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}
struct TreeNode* buildTreeHelper(int* inorder, int* postorder, int inStart, int inEnd, int* postIndex, int* inMap);struct TreeNode* buildTreeHelper(int* inorder, int* postorder, int inStart, int inEnd, int* postIndex, int* inMap) {
    if (inStart > inEnd) return NULL;
    struct TreeNode* root = newNode(postorder[(*postIndex)--]);
    int inIndex = inMap[root->val + 3000];
    root->right = buildTreeHelper(inorder, postorder, inIndex + 1, inEnd, postIndex, inMap);
    root->left = buildTreeHelper(inorder, postorder, inStart, inIndex - 1, postIndex, inMap);
    return root;
}
struct TreeNode* buildTree(int* inorder, int inorderSize, int* postorder, int postorderSize) {
    int* inMap = (int*)malloc(6001 * sizeof(int));
    for (int i = 0; i < inorderSize; i++) {
        inMap[inorder[i] + 3000] = i;
    }
    int postIndex = postorderSize - 1;
    struct TreeNode* root = buildTreeHelper(inorder, postorder, 0, inorderSize - 1, &postIndex, inMap);
    free(inMap);
    return root;
}